README.md
<a href="https://www.npmjs.com/package/@awesomeeng/awesome-log">![npm](https://img.shields.io/npm/v/@awesomeeng/awesome-log.svg "npm Details")</a>
<a href="./LICENSE">![GitHub](https://img.shields.io/github/license/awesomeeng/awesome-log.svg "License Details")</a>
<a href="http://npm-stats.com/~packages/@awesomeeng/awesome-log">![npm](https://img.shields.io/npm/dt/@awesomeeng/awesome-log.svg "npm download stats")</a>
<a href="https://github.com/awesomeeng/awesome-log/graphs/contributors">![GitHub contributors](https://img.shields.io/github/contributors-anon/awesomeeng/awesome-log.svg "Github Contributors")</a>
<a href="https://github.com/awesomeeng/awesome-log/commits/master">![GitHub last commit](https://img.shields.io/github/last-commit/awesomeeng/awesome-log.svg "Github Commit Log")</a>
<br/><a href="https://nodejs.org/en/">![node](https://img.shields.io/node/v/@awesomeeng/awesome-log.svg "NodeJS")</a>
<a href="https://github.com/awesomeeng/awesome-log/issues">![GitHub issues](https://img.shields.io/github/issues/awesomeeng/awesome-log.svg "Github Issues")</a>
<a href="https://snyk.io/vuln/search?type=npm&q=@awesomeeng/awesome-log">![Snyk Vulnerabilities for GitHub Repo](https://img.shields.io/snyk/vulnerabilities/github/awesomeeng/awesome-log.svg "Synk Vulnerabilities Database")</a>
<a href="https://www.npmjs.com/package/@awesomeeng/awesome-log">![David](https://img.shields.io/david/awesomeeng/awesome-log.svg)</a>

# AwesomeLog

AwesomeLog is a Log System for enterprise nodejs applications. It provides a basic out of the box logging solution that is ready to go with zero configuration but also gives you a highly configurable logging solution with the power to do your logging right.

## Why Another Log Solution?

AwesomeLog is similar to Winston, Log, pino, Log4js, etc. and those are all good products.  AwesomeLog just provides a different method of Logging; one we think is cleaner and more stable. If you want to use Winston/Log/pino/Log4js/whatever, that's perfectly fine by us. But if you want to try something a little cleaner, with a bit more features, consider AwesomeLog.

## Features

AwesomeLog provides...
 - Ready to go with zero initial configuration;
 - Configure and Start once, Use Everywhere;
 - Support for clean nested AwesomeLog usage;
 - Extremely flexible;
 - Customizable Log Levels;
 - Configurable log entry field contents;
 - Built-In Formatters: Default, JSON, JS, or CSV;
 - Or add your own log formatters;
 - Console and File writers;
 - or add your own custom log writer;
 - Colorized Console Logging;
 - Log History;
 - Pause/Resume;
 - SubProcess Logging;

## Contents
 - [Installation](#installation)
 - [Setup](#setup)
 - [Log Levels](#log-levels)
 - [Log Writers](#log-writers)
 - [Log Formatters](#log-formatters)
 - [Fields](#fields)
 - [Performance](#performance)
 - [Documentation](#documentation)
 - [Examples](#examples)
 - [Awesome Engineering](#the-awesome-engineering-company)
 - [Support and Help](#support-and-help)
 - [License](#license)

## Installation

Couldn't be easier.
```
npm install --save @awesomeeng/awesome-log
```

## Setup

It is best if you configure AwesomeLog at the top of your application. We recommened you do this immediately near the top of your main node entry point.

Setup has three steps:

1). Require AwesomeLog...
```
const Log = require("@awesomeeng/awesome-log");
```

2). Configure AwesomeLog...
```
// for out of the box, zero configuration
Log.init();

// or for more control
Log.init({
	// configuration options go here
});
```

3). Start AwesomeLog...
```
Log.start();
```

That's it. AwesomeLog is running and you can now use in any part of your application:

## Usage

To use AwesomeLog once it is configured and started is simple:

1). You need to require AwesomeLog in any file/class that would reference it.

```
const Log = require("@awesomeeng/awesome-log");
```

2). Then you just log stuff out.
```
Log.info("message",a,b,c);
```

Simple.

You'll notice that Log calls take one or more arguments:

 * The first argument is the log message you want to display.
 * All of the remaining arguments are extra details.  Depending on the formatter you are using, these may displayed differently. For example, if you use the `default` formatter, AwesomeLog will attempt to display printable versions of the remaining arguments; Errors in particular are formatted for easier reading.

## Log Levels

AwesomeLog starts with five basic Log Levels: ACCESS, ERROR, WARN, INFO, DEBUG by default). Each of these associated methods on the Log instance to allow you to easily use them.  So for the default levels you get
```
Log.access(...)
Log.error(...)
Log.warn(...)
Log.info(...)
Log.debug(...)
```

If you customize the log levels, AwesomeLog will remove the default log level methods (`Log.warn()` etc) and replace them with methods that have your log levels names:

```
Log.init({
	levels: "silly,goofy,funny,bad"
});
Log.start()
Log.silly(...);
Log.goofy(...);
Log.funny(...);
Log.bad(...);
```

You may have as many Log Levels as you desire, but each must have a unique name.  Also, there are a handful of reserved words (anything that is already in the AwesomeLog class) and you cannot use them. These include: `AbstractLogWriter, AbstractLogFormatter, initialized, running, config, history, historySizeLimit, levels, levelNames, definedWriters, definedFormatters, defineWriter, defineFormatter, init, start, stop, pause, resume, clearHistory, getLevel, log, captureSubProcess, releaseSubProcess.`

In addition to the log level methods, you also use the `Log.log()` method which takes the level as the first argument, so you can programmatically set the level if needed.
```
Log.log("level","message");
```

## Log Writers

A Log Writer is the channel by which a log message is written to some output device.  Writers are used to send the message to the console, the file system, a database, wherever you need your logs to end up.

AwesomeLog ships with three built-in Log Writers:

 - **Console**: Used to output log messages to the STDOUT console. Has the option to colorize output for ANSI compatible terminals.

 - **File**: Used to output log message to a given file. Has the capability to specify custom file names including date patterns, and support file rotation and cleanup.

 - **Null**: Used to output log messages to /dev/null and make them disappear forever.

By default AwesomeLog is configured to use the Console writer only. You can change the writers during your `Log.init()` call if you so desire. Here's an example:

```
Log.init({
	writers: [{
		name: "MyConsoleWriter",
		type: "console",
		formatter: "default",
		options: {
			colorize: true
		}
	},{
		name: "MyFileWriter",
		type: "file",
		formatter: "json",
		options: {
			housekeeping: "2 hours",
			filename: "logs/MyLogs.{YYYYMMDD}.log"
		}
	}],
});
```

This example enables two writers, a console writer and a file writer.

You can read more about [Configuration](./docs/Configuration), [Console Writer](./docs/ConsoleWriterConfiguration), or [File Writer](./docs/FileWriterConfiguration), or how to write your own [Custom Log Writer](./docs/LogWriters) in the respective documentation.

Each LogWriter is run in its own isolated process and log messages are passed to it. This allows AwesomeLog to execute much faster.

## Log Formatters

A Log Formatter takes the log message details (as well as a number of other system information details) and formats that into a single string message. Each Log Writer (see above) can have it's own Log Formatter.  If you do not provide a formatter, `default` is used.

AwesomeLog ships with four built-in Log Formatters:

 - **Default**: The default formatter outputs log messages in AwesomeLog's own custom format.  This includes the timestamp, the process id, the log level, the log system name, the log message, and any additional arguments.

	```
	2018-09-10T15:46:27.714Z : #1234 : INFO       : AwesomeLog.js    : Log initialized.
	2018-09-10T15:46:27.716Z : #1234 : INFO       : AwesomeLog.js    : Log started.
	```

 - **JSON**: The JSON formatter writes the entire log message out as a json string. You'll notice that there is a lot more detail in this example when compared with the one for `default` from above. Formatters often reduce the amount of log information.

	```
	{"hostname":"blah.awesomeeng.com","domain":"awesomeeng.com","servername":"blah","pid":1234,"ppid":5678,"main":"/code/project","arch":"x64","platform":"linux","bits":64,"cpus":8,"argv":"","execPath":"node","startingDirectory":"/code/project","homedir":"/home/blah","username":"blah","version":"v10.9.0","level":"INFO","system":"AwesomeLog.js","message":"Log initialized.","args":[],"timestamp":1536594573580}
	{"hostname":"blah.awesomeeng.com","domain":"awesomeeng.com","servername":"blah","pid":1234,"ppid":5678,"main":"/code/project","arch":"x64","platform":"linux","bits":64,"cpus":8,"argv":"","execPath":"node","startingDirectory":"/code/project","homedir":"/home/blah","username":"blah","version":"v10.9.0","level":"INFO","system":"AwesomeLog.js","message":"Log started.","args":[],"timestamp":1536594573582}
	```

 - **CSV**: The CSV formatter outputs the log information in a CSV format friendly to spreadsheets. The CSV format is `timestamp(unix epoch),level,pid,system name,message,arg0,arg1,...arg9999`.

 	```
	1536594995710,"INFO",19848,"AwesomeLog.js","Log initialized."
	1536594995712,"INFO",19848,"AwesomeLog.js","Log started."
	```

 - **JS**: The JS formatter outputs the log information as a JavaScript Object. It is more suited to programatic uses then to human readable usage.

	```
	[object Object]
	[object Object]
	```

You can read more about how to write your own [Custom Log Formatter](./docs/LogFormatters) in the documentation.

## Fields

AwesomeLog supplies a variety of field options.  During configuration you can specify which fields you want to be in each log entry, although what gets output is also determined by the log formatter you choose.  For example, you can specify the `hostname` field, but the Default formatter doesnt use it.  Some formatters do output the entire log entry structure though like the CSV Formatter, the JSON formatter, or the JS Object formatter.

It is worth noting that some fields are more expensive than others, performance wise. In particular, the `system` field if very expensive from a performance point of view.  If performance is a consideration for you, please consider leaving this field out.

Fields are configured thus:

```
Log.init({
	fields: "timestamp,pid,system,text,args"
});
```

The `fields` configuration property is a comma delimited string of each field name.  Order is irrelevant as the formatter you choose will dictate its own ordering.

The default fields are `timestamp,pid,system,text,args`. If your `NODE_ENV` environment variable is set to `prod` or `production` the default fields
will omit `system` to encourage better performance on production environments.

The following fields are available:
 - **timestamp**: The current time, in UTC.
 - **level**: The log level for this particular log entry.
 - **text**: The text message for this particular log entry.
 - **args**: Any extra arguments passed to the log call.
 - **system**: The javascript file in which this log call was made. Please note that including this field is very expensive from a performance standpoint, and it should only be used in development systems.
 - **hostname**: The hostname of the machine the log event occured on. Good for integrated log systems.
 - **domain**: The domain name portion of the hostname, that is the last two segments of the hostname.
 - **servername**: The computer name for the machine the log event occured on. This is the first segment of the hostname.
 - **pid**: The process id for this log message. In subprocess systems this can be a valuable debugging tool.
 - **ppid**: The parent process id.
 - **main**: The path to the executed node executable.
 - **execpath**: The path to the executed node executable.
 - **argv**: The initial arguments passed to the executed node executable.
 - **arch**: The architecture value for the current OS.
 - **platform**: The platform value for the current OS.
 - **bits**: The bits value for the current OS.
 - **cpus**: The number of physical CPUs on the current machine.
 - **startingdir**: The starting directory of the node application.
 - **homedir**: The user's home directory.
 - **username**: The user's username.
 - **version**: The version of node.

## Performance

AwesomeLog is configured by default to perform in manner best suited to development.  For production systems, a slioghtly more nuanced configuration will perform better, but at a reduction in development information. When run with this performance configuration AwesomeLog will perform slightly better than most other nodejs logging systems like Winston, Bunyan, Bole, Debug, etc. and performs about the same as pino.

#### Recommended production configuration

```
Log.init({
	fields: "timestamp,level,pid,text,args",
	separate: true,
	buffering: true,
	history: false
});
```

With regards to the `fields` configuration property: most fields are fine, however, the `system` field does add a significant performance cost and should only be used in development environments.

## Documentation

 - [Advanced Configuration](./docs/Configuration.md)
 - [Console Writer Configuration](./docs/ConsoleWriterConfiguration.md)
 - [File Writer Configuration](./docs/FileWriterConfiguration.md)
 - [Nested Usage](./docs/NestedUsage.md)
 - [Custom Log Levels](./docs/LogLevels.md)
 - [Custom Log Writers](./docs/LogWriters.md)
 - [Custom Log Formatters](./docs/LogFormatters.md)
 - [Working with Child Processes](./docs/ChildProcess.md)
 - [History](./docs/History.md)
 - [Hooks](./docs/Hooks.md)

 - [API Documentation](./docs/API.md)

## Examples

AwesomeLog ships with a set of examples for your reference.

 - [ExampleUsage](./examples/ExampleUsage): A basic example of how to use AwesomeLog.

 - [ExampleCustomWriter](./examples/ExampleCustomWriter): Illustrates how to build your own custom writer, should you need to do that.

 - [ExampleCustomFormatter](./examples/ExampleCustomFormatter): Provides a very basic example of building your own custom formatter.

 - [ExampleSubProcess](./examples/ExampleSubProcess): An example of using AwesomeLog with Child Processes.

## The Awesome Engineering Company

AwesomeLog is written and maintained by The Awesome Engineering Company. We belive in building clean, configurable, creative software for engineers and architects and customers.

To learn more about The Awesome Engineering Company and our suite of products, visit us on the web at https://awesomeeng.com.

## Support and Help

This product is maintained and supported by The Awesome Engineering Company.  For support please [file an issue](./issues) or contact us via our Webiste at [https://awesomeeng.com](https://awesomeeng.com).  We will do our best to respond to you in a timely fashion.

## License

AwesomeLog is released under the MIT License. Please read the  [LICENSE](./LICENSE) file for details.


docs/History.md
# [AwesomeLog](../README.md) > History

AwesomeLog will keep a small number of the last N log message. This is called the history and it is available for programmatic and testing purposes.

## Configuring History

The history setting are configured as part of the `Log.init()` configuration.  You can read more about [configuration](./Configuration.md) in our detailed documentation on the subject.  Specifically the following properties apply to history:

 - **history** [boolean] - Set to false if you do not want AwesomeLog to keep a small history of log messages.  The history can be used for programatic access to the log. History is enabled by default.

 - **histroySizeLimit** [number] - The number of log entries kept in the history. 1000 by default.

 - **historyFormatter** [string] - The formatter used when writing messages to the history. `default` is used by default. It should be noted that this formatter only applies to the history, not the writers.  It is often better when working with history to set this to be `js` and just get the raw log message object back.

## Accessing the History

You access the history of AwesomeLog with the `Log.history` member.  This will return an array of the last N log messages.

You also have access to the `Log.historySizeLimit` member for getting the configured size limit.

## Performance

Turning on history slow down AwesomeLog's overall performance. If this is of concern to you, please consider turning off history.


src/Log.js
// (c) 2018, The Awesome Engineering Company, https://awesomeneg.com

"use strict";

const OS = require("os");
const Events = require("events");
const Process = require("process");

const AwesomeUtils = require("AwesomeUtils");

const LogLevel = require("./LogLevel");
const LogWriter = require("./LogWriter");
const LogFormatter = require("./LogFormatter");

let Worker;
try {
	Worker = require('worker_threads');
}
catch (ex) {
	Worker = null;
}

const $CONFIG = Symbol("config");
const $DEFINED_WRITERS = Symbol("defined_writers");
const $DEFINED_FORMATTERS = Symbol("defined_formatters");
const $BACKLOG = Symbol("backlog");
const $HISTORY = Symbol("history");
const $FUNCTIONS = Symbol("functions");
const $LEVELS = Symbol("levels");
const $WRITERS = Symbol("writers");
const $RUNNING = Symbol("running");
const $SUBPROCESSES = Symbol("subprocesshandler");
const $BASE = Symbol("base");

class Log extends Events {
	constructor() {
		super();

		const hostname = OS.hostname();
		const domain = hostname.split(".").slice(-2).join(".");
		const servername = hostname.split(".").slice(0,1);
		let bits = 32;
		let arch = OS.arch();
		if (arch==="arm64" || arch==="mipsel" || arch==="ppc64" || arch==="s390x" || arch==="x64") bits = 64;

		this[$BASE] = {
			hostname,
			domain,
			servername,
			pid: process.pid,
			ppid: process.ppid,
			main: process.mainModule.filename,
			arch: OS.arch(),
			platform: OS.platform(),
			bits,
			cpus: OS.cpus().length,
			argv: process.argv.slice(2).join(" "),
			execPath: process.execPath,
			startingDirectory: process.cwd(),
			homedir: OS.homedir(),
			username: OS.userInfo().username,
			version: process.version
		};

		this[$CONFIG] = null;
		this[$DEFINED_WRITERS] = {};
		this[$DEFINED_FORMATTERS] = {};
		this[$BACKLOG] = [];
		this[$HISTORY] = [];
		this[$FUNCTIONS] = {};
		this[$LEVELS] = [];
		this[$WRITERS] = [];
		this[$RUNNING] = false;
		this[$SUBPROCESSES] = new Map();

		initLevels.call(this,"access,error,warn,info,debug");
	}

	get LogWriter() {
		return LogWriter;
	}

	get LogFormatter() {
		return LogFormatter;
	}

	get initialized() {
		return this[$CONFIG]!==null;
	}

	get running() {
		return this[$RUNNING];
	}

	get config() {
		return AwesomeUtils.Object.extend({},this[$CONFIG]);
	}

	get history() {
		return (this[$HISTORY]||[]).slice();
	}

	get historySizeLimit() {
		return this[$CONFIG].historySizeLimit;
	}

	get levels() {
		return this[$LEVELS];
	}

	get levelNames() {
		return this[$LEVELS].map((level)=>{
			return level.name;
		});
	}

	get definedWriters() {
		return AwesomeUtils.Object.extend(this[$DEFINED_WRITERS]);
	}

	get definedFormatters() {
		return AwesomeUtils.Object.extend(this[$DEFINED_FORMATTERS]);
	}

	defineFormatter(name,konstructor) {
		if (!name) throw new Error("Missing formatter name.");
		name = name.toLowerCase();

		if (!konstructor) throw new Error("Missing formatter constructor");
		if (!LogFormatter.isPrototypeOf(konstructor)) throw new Error("Invalid formatter constructor. Must inherit from LogFormatter.");

		if (this[$DEFINED_FORMATTERS][name]) throw new Error("Formatter already defined.");

		this[$DEFINED_FORMATTERS][name] = new konstructor(this);

		this.emit("formatter_added",name);
	}

	defineWriter(name,konstructor) {
		if (!name) throw new Error("Missing writer name.");
		name = name.toLowerCase();

		if (!konstructor) throw new Error("Missing writer constructor");
		if (!LogWriter.isPrototypeOf(konstructor)) throw new Error("Invalid writer constructor. Must inherit from LogWriter.");

		if (this[$DEFINED_WRITERS][name]) throw new Error("Writer already defined.");

		this[$DEFINED_WRITERS][name] = konstructor;

		this.emit("writer_added",name);
	}

	init(config) {
		if (this.initialized && this.running) throw new Error("Cannot initialize while running. stop() first.");

		let disableSP = config && config.disableSubProcesses || false;
		this[$CONFIG] = AwesomeUtils.Object.extend({
			history: true,
			historySizeLimit: 100,
			historyFormatter: "default",
			levels: "access,error,warn,info,debug",
			disableLoggingNotices: !disableSP && isSubProcess() ? true : false,
			loggingNoticesLevel: "info",
			writers: [],
			backlogSizeLimit: 1000,
			disableSubProcesses: false
		},config||{});
		disableSP = this[$CONFIG].disableSubProcesses;
		if (this[$CONFIG].writers.length<1) this[$CONFIG].writers.push({
			name: "console",
			type:  !disableSP && isSubProcess() ? "subprocess" : "default",
			levels: "*",
			formatter: !disableSP && isSubProcess() ? "subprocess" : "default",
			options: {}
		});

		initLevels.call(this,this.config.levels);
		if (!this.getLevel(this.config.loggingNoticesLevel)) this[$CONFIG].loggingNoticesLevel = this.levels.slice(-1)[0] && this.levels.slice(-1)[0].name || null;

		this[$CONFIG].historyFormatter = this[$DEFINED_FORMATTERS][this[$CONFIG].historyFormatter.toLowerCase()];
		if (!this[$CONFIG].historyFormatter) throw new Error("Invalid history formatter.");

		if (!this.config.disableLoggingNotices) this.log(this.config.loggingNoticesLevel,"AwesomeLog","Log initialized.");
		this.emit("initialized",config);

		return this;
	}

	start() {
		if (this.running) return;

		if (this[$CONFIG]===null) this.init();

		this[$RUNNING] = true;
		this[$HISTORY] = [];

		initWriters.call(this);

		[...this[$SUBPROCESSES].keys()].forEach((subprocess)=>{
			this[$SUBPROCESSES].delete(subprocess);
			this.captureSubProcess(subprocess);
		});

		if (this[$BACKLOG]) {
			this[$BACKLOG].forEach((logentry)=>{
				write.call(this,logentry);
			});
		}
		this[$BACKLOG] = null;

		if (!this.config.disableLoggingNotices) this.log(this.config.loggingNoticesLevel,"AwesomeLog","Log started.");
		this.emit("started");

		return this;
	}

	stop() {
		if (!this.running) return;
		this[$BACKLOG] = this[$BACKLOG] || [];
		this[$RUNNING] = false;

		[...this[$SUBPROCESSES].keys()].forEach((subprocess)=>{
			this.releaseSubProcess(subprocess);
			this[$SUBPROCESSES].set(subprocess,null);
		});

		this[$WRITERS].forEach((writer)=>{
			writer.flush();
			writer.close();
		});

		if (!this.config.disableLoggingNotices) this.log(this.config.loggingNoticesLevel,"AwesomeLog","Log stopped.");
		this.emit("stopped");

		return this;
	}

	pause() {
		if (!this.running) return;
		this[$BACKLOG] = this[$BACKLOG] || [];

		if (!this.config.disableLoggingNotices) this.log(this.config.loggingNoticesLevel,"AwesomeLog","Log paused.");
		this.emit("paused");

		return this;
	}

	resume() {
		if (!this.running) return;

		if (this[$BACKLOG]) {
			this[$BACKLOG].forEach((logentry)=>{
				write.call(this,logentry);
			});
		}
		this[$BACKLOG] = null;

		if (!this.config.disableLoggingNotices) this.log(this.config.loggingNoticesLevel,"AwesomeLog","Log resumed.");
		this.emit("resumed");

		return this;
	}

	clearHistory() {
		this[$HISTORY] = [];

		return this;
	}

	getLevel(level) {
		if (!level) throw new Error("Missing level argument.");
		if (level instanceof LogLevel) return level;
		if (typeof level==="string") {
			level = level.toUpperCase();
			let index = this[$LEVELS].reduce((index,x,i)=>{
				if (index>-1) return index;
				if (x && x.name && x.name===level) return i;
				return -1;
			},-1);
			return index>-1 && this[$LEVELS][index] || null;
		}
		throw new Error("Invalid level argument.");
	}

	log(level,system,message,...args) {
		let logentry = null;

		if (!system && level && typeof level==="object" && !(level instanceof LogLevel) && level.level && level.system && level.message && level.args) {
			logentry = level;
			system = logentry.system;
			message = logentry.message;
			args = logentry.args;
			level = logentry.level;
		}

		level = this.getLevel(level);
		if (!level) throw new Error("Missing level argument.");
		if (!(level instanceof LogLevel)) throw new Error("Invalid level argument.");
		if (logentry) logentry.level = level;

		if (!system) {
			system = AwesomeUtils.Module.source(1).split(/\\\\|\\|\//g).slice(-1)[0];
		}
		if (typeof system!=="string") throw new Error("Invalid system argument.");
		system = system.replace(/[^\w\d_\-.]/g,""); // strip out any non-alpha characters. _ - and . are also allowed.
		if (logentry) logentry.system = system;

		args = [].concat(args); // has to come before message check
		if (logentry) logentry.args = args;

		if (!message) throw new Error("Missing message argument.");
		if (message instanceof Error) {
			args.unshift(message);
			message = message.message;
		}
		if (typeof message!=="string") throw new Error("Invalid message argument.");

		logentry = Object.assign(this[$BASE],{
			level,
			system,
			message,
			args,
			timestamp: Date.now()
		},logentry||{});

		if (this[$BACKLOG]) {
			this[$BACKLOG].push(logentry);
			if (this[$BACKLOG].length>this.config.backlogSizeLimit) this[$BACKLOG] = this[$BACKLOG].slice(Math.floor(this.backlogSizeLimit*0.1));
		}
		else write.call(this,logentry);

		this.emit("log",logentry);

		return this;
	}

	captureSubProcess(subprocess) {
		if (!subprocess) return;
		if (!subprocess.on) return;
		if (this[$SUBPROCESSES].has(subprocess)) return;

		this[$SUBPROCESSES].set(subprocess,subProcessHandler.bind(this));
		subprocess.on("message",this[$SUBPROCESSES].get(subprocess));

		return this;
	}

	releaseSubProcess(subprocess) {
		if (!subprocess) return;
		if (!subprocess.off) return;
		if (!this[$SUBPROCESSES].has(subprocess)) return;

		subprocess.off("message",this[$SUBPROCESSES].get(subprocess));
		this[$SUBPROCESSES].delete(subprocess);

		return this;
	}
}

const isSubProcess = function isSubProcess() {
	return !!Process.channel || Worker && Worker.parentPort && Worker.parentPort.postMessage || false;
};

const initLevels = function initLevels(levels) {
	// If we have pre-existing levels, remove the functions...
	this[$LEVELS].forEach((level)=>{
		delete this[level.name.toLowerCase()];
		delete this[level.name.toUpperCase()];
		delete this[level.name.slice(0,1).toUpperCase()+level.name.slice(1).toLowerCase()];
	});

	this[$LEVELS] = [];

	let configlevels = levels;
	if (!configlevels) throw new Error("No levels configured.");
	if (typeof configlevels==="string") configlevels = configlevels.split(",");
	if (!(configlevels instanceof Array)) throw new Error("Invalid levels configured.");

	// build our levels array.
	configlevels = AwesomeUtils.Array.compact(configlevels.map((level)=>{
		if (!level) return null;
		if (level instanceof LogLevel) return level;
		if (typeof level==="string") return new LogLevel(level);
	}));
	// make sure not reserved.
	configlevels.forEach((level)=>{
		if (this[level.name.toLowerCase()] || this[level.name.toUpperCase()] ||
		this[level.name.slice(0,1).toUpperCase()+level.name.slice(1).toLowerCase()]) throw new Error("Invalid log level: '"+level.name+"' is a reserved word.");
	});

	this[$LEVELS] = configlevels;

	// add our level functions
	this[$LEVELS].forEach((level)=>{
		let lf = this.log.bind(this,level);

		this[level.name.toLowerCase()] = lf;
		this[level.name.toUpperCase()] = lf;
		this[level.name.slice(0,1).toUpperCase()+level.name.slice(1).toLowerCase()] = lf;
	});
};

const initWriters = function initWriters() {
	// clean up old writers.
	this[$WRITERS].forEach((writer)=>{
		writer.flush();
		writer.close();
	});
	this[$WRITERS] = [];

	let configwriters = this.config.writers;
	if (!configwriters) throw new Error("No writers configured.");
	if (!(configwriters instanceof Array)) throw new Error("Invalid writers configured.");

	// start new writers.
	this[$WRITERS] = AwesomeUtils.Array.compact(configwriters.map((writer)=>{
		if (!writer) return null;

		let name = writer.name || null;
		if (!name) throw new Error("Missing writer name.");
		name = name.replace(/[^\w\d_]/g,""); // strip out any non variables friendly characters.

		let type = writer.type || null;
		if (!type) throw new Error("Missing writer type.");
		type = type.toLowerCase();

		let levels = writer.levels || "*";
		if (!levels) throw new Error("Missing writer levels.");
		levels = levels.toLowerCase();

		let formatter = writer.formatter || "default";
		if (typeof formatter==="string") formatter = this[$DEFINED_FORMATTERS][formatter.toLowerCase()];
		if (!formatter) throw new Error("Missing writer formatter '"+writer.formatter+"'.");
		if (!(formatter instanceof LogFormatter)) throw new Error("Invalid writer formatter '"+writer.formatter+"', must be of type LogFormatter.");

		let options = writer.options || {};

		let konstructor = this[$DEFINED_WRITERS][type] || null;
		if (!konstructor) throw new Error("Invalid writer type '"+type+"' does not exist.");

		let instance = new konstructor(this,name,levels,formatter,options);
		return instance;
	}));
};

const write = function write(logentry) {
	if (!logentry) throw new Error("Missing log entry argument.");
	if (!logentry.level || !logentry.system || !logentry.message || !logentry.args || !logentry.timestamp) throw new Error("Invalid log entry argument.");

	if (this.config.history) {
		this[$HISTORY].push(this.config.historyFormatter.format(logentry));
		if (this.history.length> this.config.historySizeLimit) this[$HISTORY] = this[$HISTORY].slice(-this.config.historySizeLimit);
	}

	this[$WRITERS].forEach((writer)=>{
		if (!writer.takesLevel(logentry.level)) return;
		writer.write(writer.format(logentry),logentry); // message,logentry
	});
};

const subProcessHandler = function subProcessHandler(message) {
	if (!message) return;
	if (!message.cmd) return;
	if (!message.cmd==="AwesomeLog") return;

	let logentry = message.logentry;
	this.getLevel(logentry.level); // cuases an exception of the process used a level we dont know.

	this.log(logentry);
};

// create our singleton instance
let instance = new Log();

// define built in writers
instance.defineWriter("null",require("./writers/NullWriter"));
instance.defineWriter("nullwriter",require("./writers/NullWriter"));
instance.defineWriter("subprocess",require("./writers/SubProcessWriter"));
instance.defineWriter("default",require("./writers/ConsoleWriter"));
instance.defineWriter("console",require("./writers/ConsoleWriter"));
instance.defineWriter("consolewriter",require("./writers/ConsoleWriter"));
instance.defineWriter("stdout",require("./writers/ConsoleWriter"));
instance.defineWriter("file",require("./writers/FileWriter"));
instance.defineWriter("filewriter",require("./writers/FileWriter"));

// define built in formatters
instance.defineFormatter("default",require("./formatters/DefaultFormatter"));
instance.defineFormatter("subprocess",require("./formatters/SubProcessFormatter"));
instance.defineFormatter("json",require("./formatters/JSONFormatter"));
instance.defineFormatter("js",require("./formatters/JSObjectFormatter"));
instance.defineFormatter("jsobject",require("./formatters/JSObjectFormatter"));
instance.defineFormatter("csv",require("./formatters/CSVFormatter"));

// export the instance.
module.exports = instance;
